src/rosdiscover/recover/tool.py

```diff
@@ -1,20 +1,331 @@
+        ValueError
+            if the workspace does not provide a .built_by file inside its build directory
+        """
+        try:
+            package = self._app.description.packages[package_name]
+        except KeyError as err:
+            raise ValueError(f"no package found with given name: {package_name}") from err
+
+        # ensure that no absolute paths are given
+        if any(os.path.isabs(path) for path in sources):
+            raise ValueError("expected source paths to be relative to package directory")
+
+        # compute the absolute paths of each source file
+        sources = [os.path.join(package.path, path) for path in sources]
+
+        compile_commands_path = self._find_compile_commands_file(package)
+
+        self._recover(compile_commands_path, sources)
+
+    def _recover(
+        self,
+        compile_commands_path: str,
+        source_file_abs_paths: t.Collection[str],
+    ) -> None:
+        """Invokes the C++ recovery binary to recover the dynamic architecture of a given node.
+
+        Parameters
+        ----------
+        compile_commands_path: str
+            The absolute path to the compile_commands.json associated with the given node.
+        source_file_abs_paths: str
+            A list of the C++ translation unit source files (i.e., .cpp files)
+            for the given node, provided as absolute paths within the container
+        """
+        if not self._app_instance:
+            raise ValueError("tool has not been started")
+
+        logger.debug("beginning static recovery process")
+        shell = self._app_instance.shell
+        files = self._app_instance.files
+
+        if not source_file_abs_paths:
+            raise ValueError("expected at least one source file")
+
+        if not os.path.isabs(compile_commands_path):
+            raise ValueError(f"expected absolute compile commands path: {compile_commands_path}")
+
+        if not files.exists(compile_commands_path):
+            raise ValueError(f"compile_commands.json not found at given location: {compile_commands_path}")
+
+        for source_file in source_file_abs_paths:
+            if not os.path.isabs(source_file):
+                raise ValueError(f"expected absolute source file path: {source_file}")
+            if not files.exists(source_file):
+                raise ValueError(f"source file was not found: {source_file}")
+
+        for source_file in source_file_abs_paths:
+            self._prepare_source_file(source_file)
 
-    def recover(self, node_type: str, package: str) -> RecoveredNodeModel:
-        raise NotImplementedError
+        env = {
+            "PATH": "/opt/rosdiscover/bin:${PATH:-}",
+            "LIBRARY_PATH": "/opt/rosdiscover/lib:${LIBRARY_PATH:-}",
+            "LD_LIBRARY_PATH": "/opt/rosdiscover/lib:${LD_LIBRARY_PATH:-}",
+        }
+        env_args = [f"{var}={val}" for (var, val) in env.items()]
+        args = env_args + [
```

Comment:
Not sure what `rosdiscover -p` does. Maybe a comment about this would help. (e.g., I am thinking that it is calling the cli of this code, but it doesn't seem to be doing that.)

Reply:
The binary name is a little misleading. I've renamed this to `rosdiscover-cxx-extract`. This line is essentially calling the static recovery binary inside the container. `-p` specifies the directory that holds the `compile_commands.json` file.